analytics/client.go

```diff
@@ -0,0 +1,318 @@
+func NewClient(config Config) Client {
+	if config.Disabled {
+		log.Debugf("Analytics opt-out is set, analytics will be disabled")
+		return &noopClient{}
+	}
+	client := &client{
+		userID: config.UserID,
+	}
+	// Periodically flush queued analytics hits
+	ticker := time.NewTicker(1 * time.Minute)
+	go func() {
+		for {
+			<-ticker.C
+			client.Flush()
+		}
+	}()
+	return client
+}
+
+const maxHits = 20
+
+type client struct {
+	userID     uuid.UUID
+	queueLock  sync.Mutex
+	queuedHits []Params
+}
+
+func (c *client) Screenview(name string, params Params) error {
+	s, err := newScreenview(name, params)
+	if err != nil {
+		return err
+	}
+	log.Debugf("%v received", s)
+	c.enqueue(s.ToHit())
+	return nil
+}
+
+func (c *client) Event(category string, action string, params Params) error {
+	e, err := newEvent(category, action, params)
+	if err != nil {
+		return err
+	}
+	log.Debugf("%v received", e)
+	c.enqueue(e.ToHit())
+	return nil
+}
+
+func (c *client) Flush() {
+	c.queueLock.Lock()
+	defer c.queueLock.Unlock()
+	c.flush()
+}
+
+func (c *client) enqueue(hit Params) {
+	c.queueLock.Lock()
+	defer c.queueLock.Unlock()
+	if len(c.queuedHits) >= maxHits {
+		c.flush()
+	}
+	c.queuedHits = append(c.queuedHits, hit)
+}
+
+func (c *client) flush() {
+	if len(c.queuedHits) <= 0 {
+		return
+	}
+	// According to https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#batch,
+	// each line in the batch request's body represents a single hit. Thus, we log
+	// something like
+	//     Submitting analytics... (<base_params>)
+	//     Payload:
+	//       <first_hit>
+	//       ...
+	//       <last_hit>
+	// before submitting the data.
+	baseParams := c.baseParams()
+	var logMsg strings.Builder
+	logMsg.WriteString(fmt.Sprintf("Submitting analytics... (%v)\nPayload:\n", baseParams))
+	var payload []string
+	for _, hit := range c.queuedHits {
+		payload = append(payload, hit.merge(baseParams).encode())
+		logMsg.WriteString(fmt.Sprintf("  %v\n", hit))
+	}
+	log.Debug(logMsg.String())
+	body := strings.Join(payload, "\n")
+	// The Measurement Protocol's docs indicate that the endpoint
+	// will always return a 200 OK status, even if the request
+	// contains any errors. Thus, the response is useless so it is
+	// OK for us to ignore it.
+	_, err := httpClient.post(
+		"https://www.google-analytics.com/batch",
+		"application/x-www-form-urlencoded",
+		strings.NewReader(body),
+	)
+	if err != nil {
+		log.Infof("Failed to send analytics: %v", err)
+	}
+	c.queuedHits = c.queuedHits[:0]
+}
+
+func (c *client) baseParams() Params {
+	params := Params{
+		// v => Protocol Version
+		"v": "1",
+		// cid => Client ID
+		"cid": c.userID.String(),
+		// tid => Tracking ID
+		//
+		// TODO: Set this to the main Wash GA account's tracking ID.
+		"tid": "UA-144659575-2",
+		// an => Application Name
+		"an": "Wash",
+		// av => Application Version
+		//
+		// TODO: Fill this in with a proper value
+		"av": "1.0",
```

Comment:
This is a bit hard because the version's set in `cmd.version`. Wondering if this might be worth a separate version package. Maybe properly setting the version could be a separate issue?

Reply:
Reporting application version is pretty useful. A separate package that we update makes sense.

Reply:
Should that be a separate issue or should I tack it on here? Think that may need some changes to homebrew-puppet.

Reply:
I see it as part of this issue.